### ISSUE TITLE
atari8: invalid xex produced

### ISSUE THREAD

**Issue by mrk-its** [llvm-mos/llvm-mos#182]
It looks like sometimes wrong atari xex files are produced.
Attached simple program with empty main and resulting xex & elf file.

It seems end address of main section is 1 byte too big (start = 0x2000  end = 0x2016 (inclusive), section data size: 0x16 bytes)

end address in OUTPUT_FORMAT looks ok:
```
SHORT(__data_end - 1)
```
value of __data_end symbol in elf file also looks good:

```
00002016 g       .data  00000000 __data_end
```

but `0x2016` lands in binary.
So it looks like result of `__data_end - 1` expression is for some reason aligned to 2 bytes

[atari.zip](https://github.com/llvm-mos/llvm-mos-sdk/files/8645828/atari.zip)

**Comment by mysterymath** [comment]
I did some preliminary digging, and it looks like we're hitting a strange and extremely specific case in LLD.
When we're computing the value of __data_end - 1, this gets resolved down to an expression meaning: "The offset of -1 relative to the beginning of the .data section." There's a specific check for an offset of negative one; apparently this means "the size of the section." Since this section size is zero, the whole expression evaluates to __data_end, not __data_end - 1 as desired.

I'm going to have to do some more digging to find out who introduced this behavior and why; it seems bizarre, so it may be a backwards compatibility thing or something. I'll also need to figure out what to do about this.

**Comment by mrk-its** [comment]
Thanks for taking a look on that so quickly. Knowing all above - do you think there is some temporary workaround possible to fix `atari8` platform?

**Comment by mysterymath** [comment]
It looks like the -1 was added as a hack for a specific purpose, but it doesn't look like it actually uses the mechanism anymore. I've removed the check in LLD. Fixed by 28dc874d.

**Comment by mysterymath** [comment]
The fix actually broke quite a few LLD tests; I'm going to try and do a more principled solution to this. Hopefully something I can upstream later.

**Comment by mysterymath** [comment]
Fixed by llvm-mos/llvm-mos-sdk@255e2a9
